### ISSUE TITLE
Automatically change compat entries on Pkg.update

### ISSUE THREAD

**Issue by fonsp** [JuliaLang/Pkg.jl#4092]
https://github.com/JuliaLang/Pkg.jl/pull/3732 automatically adds compat entries when you add a package (to a named environment). The compat entry looks like:

```toml
[compat]
Example = "1.2.3"
```

My proposal is that compat entries should be changed automatically when you `Pkg.update()`. E.g. when you update to `1.3.0`, the Project.toml should change to:

```toml
[compat]
Example = "1.3.0"
```

I believe that the same reasoning behind https://github.com/JuliaLang/Pkg.jl/pull/3732 applies to the `update` situation.

(This same behaviour also happens in `yarn` and `npm`, where the `install` adds `^1.2.3` to the package.json, and `upgrade` changes it to `^1.3.0`.)

**Comment by KristofferC** [comment]
This doesn't make sense to me? Why would you want to claim incompatible with previous versions just because you do an `update`?

**Comment by fonsp** [comment]
Thanks for the comments! To me it makes sense as a default, because `update` changes the version your are developing and testing with, and you might start depending on new functionality that was added in a minor update. 

So as a "safe" default, `1.3.0` seems better after updating. To me it feels like the same argument for using `1.2.3` as the default for `add`, instead of just `1`.

But I also see some practical issues like the one @jakobnissen brought up, where it is useful in the Julia ecosystem for packages to have wide compat ranges and we don't want to override those with stricter ones. `npm` is different here, since dependencies are not necessarily shared (two packages in one project can import different versions of the same transitive dep), so restricting compat bounds upwards is never really an issue there.

**Comment by okartal** [comment]
> But then, if the author _did_ check it was compatible with `Example 1.2.3`, wouldn't it be annoying that the compat was incorrectly changed?

I disagree. Why is it annoying to **explicitly** state and update your dependencies? An update can always introduce a buggy version of some dependency which may pop up only later. I can never _check_ exhaustively if my package is compatible. At the moment, the only way to track the exact dependency of your package is to track the Manifest.toml - this is, arguably, even more annoying.

**Comment by jakobnissen** [comment]
@okartal This issue is about automatically removing support for older versions of a dependency, even if this older version is tested for.
It's not about specifying the specific versions the developer tested it with - that's also rarely useful, because the user almost certainly will need to use it with a different combination of packages.
It's also not about limiting support for future versions of your dependencies.

That is going to cause _more_ bugs, not less, because it creates needless ecosystem churn.

**Comment by okartal** [comment]
@jakobnissen, you say

> It's not about specifying the specific versions the developer tested it with - that's also rarely useful, because the user almost certainly will need to use it with a different combination of packages

This may be true, if you create a package that is used from within Julia alongside other packages you have no control over - but if you create a library and want to compile it to be used in software, you want to pin the exact versions you tested your library with. In enterprise software, this is mandatory, I'd say.

**Comment by KristofferC** [comment]
I don't think this will be done.
